lib/shape.ml

```diff
@@ -60,6 +51,72 @@ let line ?x1 ?y1 x2 y2 =
   | Some x, Some y -> Line { a = { x; y }; b = { x = x2; y = y2 } }
   | _ -> Line { a = { x = 0; y = 0 }; b = { x = x2; y = y2 } }
 
+let polygon lst_points = Polygon lst_points
+
+let rec take n lst =
+  match (n, lst) with
+  | 0, _ -> ([], lst)
+  | _, [] -> ([], [])
+  | n, x :: xs ->
+      let taken, rest = take (n - 1) xs in
+      (x :: taken, rest)
+
+let rec partition n step lst =
+  match lst with
+  | [] -> []
+  | _ ->
+      let taken, _ = take n lst in
+      if List.length taken = n then taken :: partition n step (List.tl lst)
+      else []
+
+let render_polygon polygon =
+  let denorm_points = List.map denormalize polygon in
+  let point_to_tuples =
+    List.map (fun { x : int; y : int } -> (x, y)) denorm_points
+  in
+  let partitioned =
+    partition 2 1 (point_to_tuples @ [ List.hd point_to_tuples ])
+  in
+  let array_of_points = Array.of_list (List.flatten partitioned) in
+  draw_poly_line array_of_points
+
+let render_rect rect =
+  match rect with
+  | Rectangle rectangle ->
+      let points =
+        [
+          rectangle.c;
+          { x = rectangle.c.x; y = rectangle.c.y + rectangle.length };
+          {
+            x = rectangle.c.x + rectangle.width;
+            y = rectangle.c.y + rectangle.length;
+          };
+          {
+            x = rectangle.c.x + rectangle.width;
+            y = rectangle.c.y + rectangle.length;
+          };
+        ]
+      in
+      render_polygon points
+  | _ -> ()
```

Comment:
we should get rid of the previous `Rectangle` constructor entirely.

the problem is, e.g., we can't store the rotation with the amount of data our rectangle type currently has. we need to change that data representation entirely.

the `rectangle` constructor can still take the same arguments and calculate the polygon points accordingly.